crates/flipperzero/src/furi/stream_buffer.rs

```diff
@@ -0,0 +1,500 @@
+    ///
+    /// The `timeout` is ignored when called from an interrupt routine.
+    pub unsafe fn receive(&self, data: &mut [u8], timeout: furi::time::Duration) -> usize {
+        let self_ptr = self.0.as_ptr();
+        let data_ptr: *mut c_void = data.as_mut_ptr().cast();
+        let data_len = data.len();
+        let timeout = timeout.0;
+        unsafe { sys::furi_stream_buffer_receive(self_ptr, data_ptr, data_len, timeout) }
+    }
+
+    /// Returns the number of bytes currently available in the buffer.
+    pub fn bytes_available(&self) -> usize {
+        let self_ptr = self.0.as_ptr();
+        unsafe { sys::furi_stream_buffer_bytes_available(self_ptr) }
+    }
+
+    /// Returns the number of bytes that can still fit in the buffer.
+    pub fn spaces_available(&self) -> usize {
+        let self_ptr = self.0.as_ptr();
+        unsafe { sys::furi_stream_buffer_spaces_available(self_ptr) }
+    }
+
+    /// Checks if the buffer is full.
+    pub fn is_full(&self) -> bool {
+        let self_ptr = self.0.as_ptr();
+        unsafe { sys::furi_stream_buffer_is_full(self_ptr) }
+    }
+
+    /// Checks if the buffer is empty.
+    pub fn is_empty(&self) -> bool {
+        let self_ptr = self.0.as_ptr();
+        unsafe { sys::furi_stream_buffer_is_empty(self_ptr) }
+    }
+
+    /// Attempts to reset the stream buffer.
+    ///
+    /// Clears the buffer, discarding any data it contains and returning it to its initial empty
+    /// state.
+    /// The reset can only succeed if no tasks are blocked waiting to send or receive data;
+    /// otherwise, an [`Err`] is returned.
+    pub fn reset(&self) -> furi::Result<()> {
+        let status = unsafe { sys::furi_stream_buffer_reset(self.0.as_ptr()) };
+        let status = Status(status);
+        status.err_or(())
+    }
+}
+
+impl Drop for StreamBuffer {
+    fn drop(&mut self) {
+        unsafe {
+            sys::furi_stream_buffer_free(self.0.as_ptr());
+        }
+    }
+}
+
+/// Error when setting a too large [trigger level](StreamBuffer::set_trigger_level) on
+/// [`StreamBuffer`].
+#[derive(Debug, uDebug)]
+pub struct SetTriggerLevelError;
+
+impl Display for SetTriggerLevelError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "specified trigger level exceeds the buffer's length")
+    }
+}
+
+impl uDisplay for SetTriggerLevelError {
+    fn fmt<W>(&self, f: &mut ufmt::Formatter<'_, W>) -> Result<(), W::Error>
+    where
+        W: ufmt::uWrite + ?Sized,
+    {
+        uwrite!(f, "specified trigger level exceeds the buffer's length")
+    }
+}
+
+impl core::error::Error for SetTriggerLevelError {}
+
+#[cfg(feature = "alloc")]
+pub use stream::*;
+
+#[cfg(feature = "alloc")]
+mod stream {
+    use crate::furi;
+
+    use super::*;
+
+    use alloc::sync::Arc;
+    use core::{cell::Cell, marker::PhantomData};
+
+    /// A zero-sized type used to mark types as `!Sync`.
+    type PhantomUnsync = PhantomData<Cell<()>>;
```

Comment:
For future work on our side: this should probably be moved to some internal `utils` module